### PR TITLE
feat: different rate limiting handling

### PR DESCRIPTION
**Pull request by pauldambra**
the last attempt didn't play nicely with older clients...

Now

* capture still returns 200 when rate limiting but includes a list of rate limited products in the body
* the rate limiter uses the batch key or defaults to events (currently recordings sets a batch key and events doesn't)
* we rate limit for 1 minute
* we don't check for rate limiting on every response since there's no need to parse json for every request just in case there's been limiting

### CODE REVIEW

File: src/extensions/sessionrecording.ts [hunk 1]
```diff
@@ -331,7 +331,7 @@ export class SessionRecording {
 
         this.mutationRateLimiter =
             this.mutationRateLimiter ??
-            new MutationRateLimiter(this.rrwebRecord!, {
+            new MutationRateLimiter(this.rrwebRecord, {
```
Comment (pauldambra) [inline src/extensions/sessionrecording.ts:334]: this is checked above so doesn't need the not null assertion

File: src/send-request.ts [hunk 2]
```diff
@@ -113,8 +114,8 @@ export const xhr = ({
                     onXHRError(req)
                 }
 
-                // don't retry certain errors
-                if ([401, 403, 404, 500].indexOf(req.status) < 0) {
+                // don't retry errors between 400 and 500 inclusive
```
Comment (pauldambra) [inline src/send-request.ts:117]: sneaky fly-by - I think this is neater but 🤷

File: src/rate-limiter.ts [hunk 3]
```diff
@@ -26,26 +24,27 @@ export class RateLimiter {
         return new Date().getTime() < retryAfter
     }
 
-    on429Response(response: XMLHttpRequest): void {
-        if (response.status !== 429) {
+    // this needs to be an arrow function so that it can be passed as a callback
+    // and have the correct `this` context in order to read `checkThreshold`
+    public checkForLimiting = (xmlHttpRequest: XMLHttpRequest): void => {
+        if (Math.random() >= this.checkThreshold) {
+            // we don't need to check this on every request
```
Comment (benjackwhite) [inline src/rate-limiter.ts:31]: I find this a little confusing. Why do we only need to check now and then? 
Feels like a step that could be confusing to us in the future when we still see a lot of requests coming in.

Is the subsequent code super slow or something? Seems like fairly straightforward code so I wouldn't say the perf benefits outweigh the extra complexity personally

File: src/rate-limiter.ts [hunk 4]
```diff
@@ -26,26 +24,27 @@ export class RateLimiter {
...
-            }
-        })
+        let response: CaptureResponse
+        try {
+            response = JSON.parse(xmlHttpRequest.responseText)
+            const quotaLimitedProducts = response.quota_limited || []
+            quotaLimitedProducts.forEach((batchKey) => {
+                if (Config.DEBUG) {
+                    console.warn(`[PostHog RateLimiter] ${batchKey || 'events'} is quota limited.`)
+                }
```
Comment (benjackwhite) [inline src/rate-limiter.ts:42]: Don't we have a logger util for exactly this?

**Review by benjackwhite — CHANGES_REQUESTED** [review]
Looks good! Small comments but the core looks great

**Review by benjackwhite — APPROVED** [review]
👍

**Review by benjackwhite — APPROVED** [review]

**Pull request merged**